### PR TITLE
Very WIP: New optimizer

### PR DESCRIPTION
**Pull request by Keno**
This PR introduces a new IR form and optimizer to address some of the performance problems encountered in #25261. The IR is pure SSA form and borrows from [B3](https://webkit.org/docs/b3/intermediate-representation.html) and LLVM in its design choices. I'm planning to write a comprehensive design document once this is a bit further along. The new IR contains Phi nodes, so a significant part of this is adding support for those in codegen and the optimizer.

The current status is that this bootstraps (in the sense of building itself) fine on top of an existing system image, and can bootstrap a new system image to the same place as doing an external bootstrap on master (somewhere around version.jl at which point it gets confused about being not-quite-Base). It also passes `core` and `string` tests, in the externally bootstrapped version. It does not however, bootstrap itself internally yet (keeps recursing into inference), so this is quite an early state. Opening this PR to give @vtjnash and @JeffBezanson a place to collaborate on this.

Once this is done, we should be able to delete most of the existing passes from optimize.jl, so I'm hoping that this'll be overall net negative.

TODO status:
- ~turning on optimization passes~
- ~line number info~ DONE
- ~handling meta nodes correctly~ DONE
- ~possible improvements (like not adding edges from `error()::Undef{}`)~
- replace the naive algorithms with the fast ones / performance
- handling enter/exit
- cleaning up code / removing debug spew

### CODE REVIEW

**Comment by vtjnash** [comment]
From discussion with Keno, we're thinking of merging this soon behind a feature-flag. He thought he might have time today to remove all of the debug statements, re-enable the old optimization passes, and squash/rebase the PR.

**Comment by Keno** [comment]
Rebased, removed disabled optimization passes, squashed, flag added to inference to turn on the new optimizer.

File: base/client.jl [hunk 1]
```diff
@@ -298,11 +298,6 @@ function exec_options(opts)
     # load ~/.julia/config/startup.jl file
     startup && load_julia_startup()
 
-    if repl || is_interactive
```
Comment (KristofferC) [inline base/client.jl]: Unrelated?

**Comment by KristofferC** [comment]
Worth to run nanosoldier? Just to make sure.

File: base/client.jl (hunk 1 above)
Comment (Keno) [reply base/client.jl]: Good catch. I assume this was broken at some point, so carelessly deleted.

**Comment by Keno** [comment]
Sure, it is disabled by default, so in theory, we should be fine, but who knows:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/26079#issuecomment-368344423) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/c19de266982fe0071099f3980362e6063d05df49/cfe7131_vs_7b2b68d/report.md). cc @ararslan

**Comment by vtjnash** [comment]
The tests compiler and cmdlineargs needs some minor fixup, then this seems good to go.

**Comment by Keno** [comment]
FreeBSD and OS X hit their respective intermittent kernel bugs. AppVeyor is as on master. Other tests went through though. Merging. As a reminder, this is just getting the code in, the actual functionality is still behind an (off by default) flag.

**Comment by StefanKarpinski** [comment]
Implicated in dot-call global scope slow down: https://github.com/JuliaLang/julia/issues/26767. I thought this was hidden behind a feature flag—is there somewhat it may affect performance in global scope even when the feature flag is turned off? If so, can we mitigate that somehow?
